### DISCUSSION TITLE
Reproduced GPT-2 (124M) + RoPE + SwiGLU — trained from scratch on FineWeb-Edu (10BT) using RTX 5090

### DISCUSSION THREAD

**Discussion by 50s-zerotohero** [karpathy/build-nanogpt#107] [Show and tell]
Hi everyone,
I’d like to share my recent project that I built while following Karpathy’s Zero-to-Hero series.
I fully reproduced GPT-2 (124M) from scratch and extended it with several modern components:

- RoPE (rotary positional embeddings, LLaMA-style)
- SwiGLU activation
- FineWeb-Edu 10BT as the training dataset
- HellaSwag evaluation (1000 samples)

🔧 Training setup
- GPU: RTX 5090 (32GB) × 1
- Training time: ~28 hours
- Framework: PyTorch
- Base: Karpathy’s "build-nanogpt" codebase

📈 Results

- Final validation loss: 3.18
- Best validation loss: 2.89
- HellaSwag accuracy (final): 0.355
- HellaSwag accuracy (best): 0.369

For reference:
- OpenAI GPT-2 (124M)	~0.292
- OpenAI GPT-3 (124M-equivalent)	~0.338
- This model (best)	0.369
➡ My model surpassed the GPT-3 (124M) baseline, likely thanks to RoPE, SwiGLU, and the dataset.

📊 Training curves

<img src="https://raw.githubusercontent.com/hideya-matsuda/nanoGPT-124M-RoPE-SwiGLU/main/plots/loss_hellaswag.png" width="700">

🔗 GitHub repository
https://github.com/hideya-matsuda/nanoGPT-124M-RoPE-SwiGLU

This project is the culmination of my entire learning journey from bi-gram → makemore → tiny transformer → GPT-2 reproduction.
If anyone wants to try reproducing the model or discuss improvements, feel free to reach out!

**Comment by durb-afk** [comment]
Hey this is soo awesome would love to discuss on how to reproduce this and improve scale and performance can you reach out to me druvc01@gmail.com or use any of the links in my profile to send me a message thanks!!

**Comment by MarkoKarbevski** [comment]
Hey, have you tried different LRs for this setup? Also do you know what loss is achieved by the Karpathy nanogpt with the same exact number of steps?

**Reply by 50s-zerotohero** [reply]
Thanks for the questions! I experimented with different values for max_lr, min_lr, and warmup_steps. However, changing the learning-rate settings did not produce a meaningful improvement, so the only change I kept was increasing warmup_steps from 715 to 750. 
I may have tested nanoGPT with the same number of training steps, but I don’t remember the exact result.
